File: app/src/main/java/com/hustunique/morii/database/DiaryDao.java
package com.hustunique.morii.database;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;

import java.util.List;

@Dao
public interface DiaryDao {

    @Insert
    long insertDiaryInfo(DiaryInfo info);

    @Insert
    void insertSoundItemInfo(SoundItemInfo info);

    @Transaction
    @Query("select * from DiaryInfo")
    List<DiaryWithSoundItemInfo> getAllDiaryWithSoundItemInfo();

    @Transaction
    default void deleteInfoById(long diaryId) {
        deleteDiaryInfoById(diaryId);
        deleteSoundInfoByDiaryId(diaryId);
    }

    @Query("delete from DiaryInfo where id=:diaryId")
    void deleteDiaryInfoById(long diaryId);

    @Query("delete from SoundItemInfo where diaryInfoId=:diaryId")
    void deleteSoundInfoByDiaryId(long diaryId);
}

File: app/src/main/java/com/hustunique/morii/database/DiaryDao.kt
package com.hustunique.morii.database

import androidx.room.*

@Dao
interface DiaryDao {
    @Insert
    fun insertDiaryInfo(info: DiaryInfo?): Long

    @Insert
    fun insertSoundItemInfo(info: SoundItemInfo?)

    @get:Query("select * from DiaryInfo")
    @get:Transaction
    val allDiaryWithSoundItemInfo: List<DiaryWithSoundItemInfo>

    @Transaction
    fun deleteInfoById(diaryId: Long) {
        deleteDiaryInfoById(diaryId)
        deleteSoundInfoByDiaryId(diaryId)
    }

    @Query("delete from DiaryInfo where id=:diaryId")
    fun deleteDiaryInfoById(diaryId: Long)

    @Query("delete from SoundItemInfo where diaryInfoId=:diaryId")
    fun deleteSoundInfoByDiaryId(diaryId: Long)
}